### PR TITLE
[Frontend] Feature/UI-CLP

### PR DESCRIPTION
**Pull request by ryardley**
_UPDATE: So we are drawing a stick in the sand for this long running PR and are looking to merge it as swap works against a local sifnode and liquidity functionality is roughly fleshed out. This will aid us in ensuring we have it integrated correctly moving forward. Remaining things in this list should be added as cards and supplied as smaller focused PRs_

Tasks done in this PR

- [x] Create VSCode project
- [x] Connect to wallet flow
- [x] Display select tokens
- [x] Fledgling Swap flow (incomplete)
- [X] Refactor Domain
- [x] Take advantage of Asset lookup and stop creating Assets everywhere
- [x] Create good display components for tokens
- [x] Attempt to setup mainnet list of tokens
- [x] Create token ingestion
- [x] Ensure tokens have images
- [x] Rough styling
- [x] Add some business logic to swap interface
- [x] Basic Swap functionality
- [x] Basic Pool functionality
- [x] Get Max button working

# Screens
- [x] Confirm Swap Form/Dialog
- [x] Confirming Dialog 
- [x] Confirmed Dialog 

Video (outdated): https://sifchain.slack.com/archives/G01BX71QY1F/p1603630398275400

# Acceptance criteria

## Query List of Available Tokens Acceptance Criterion
- [x] Click on token select display list 
  - [x] wallet balances
  - [x] [Curated list of ERC-20 tokens](https://github.com/Sifchain/sifnode/blob/feature/add-rough-interface/ui/core/data/erc20.json) generated with [this hacky script](https://github.com/Sifchain/sifnode/blob/feature/add-rough-interface/ui/core/scripts/getTopErcTokens.js) that pulls data from coingecko
- [x] User sees search bar to search for tokens.
- [x] Can search the token on the search bar
- [x] Limit to 20 erc-20 tokens 
- [x] Limit to 10 tokens from wallet
- [x] Allow user to search for unlisted tokens

## Connect to wallet Acceptance Criterion
- [x] User is presented several ways to connect
- [x] Can see token balances and connect button goes away
- [x] Disconnecting wallet cannot see balances and connect button becomes active

### 1. Appearance
- [x] From token amount
- [x] From token
- [x] To token amount
- [x] To token
- [x] ~Maximum sold (Need help)~ not required for mvp
- [x] ~Price Impact (Need help)~ not required for mvp
- [x] ~Liquidity provider fee (Need help)~ not required for mvp

### 2. Calculation
- [x] From Token
- [x] From Amount 
- [x] To Token 
- [x] To Amount 
- [x] ~Maximum sold (Need help)~ not required for mvp
- [x] ~Price Impact (Need help)~ not required for mvp
- [x] ~Liquidity provider fee (Need help)~ not required for mvp

### 3. Successful initiation
- [x] Acknowledge transaction was initiated

### 4. Transaction Success Acknowledgement
- [x] Acknowledge asynchronously successful transaction 

### 5. Transaction Failure Acknowledgement
- [ ] Acknowledge asynchronously transaction failure

### 6. Backend Update
- [ ] Market swap rate updated by the backend (Code setup to do this reasonably easily but need info on exactly what API provides liquidity pool volume to calculate rate or stream interface etc.)

### 7. Insufficient Funds
- [x] Show error message on insufficient funds
- [x] Swap button is not visible

### 8. No wallet connected
- [x] Connect wallet button is visible
- [x] Swap button is not visible
- [x] Calculate To token amount
- [x] ~Maximum sold (Need help)~ not required for mvp
- [x] ~Price Impact (Need help)~ not required for mvp
- [x] ~Liquidity provider fee (Need help)~ not required for mvp

### 1. Navigate via "Create Pool" Button
- [x] Token A & Token B
- [x] Token A Amount & Token B Amount
- [x] Can see Token Amounts they enter
- [x] Calculate Token A per B ratio
- [x] Calculate Token B per A ratio
- [x] Calculate Share of Pool

### 2. Enter token values less than wallet values
- [x] Calculate: Amount of Token A that user owns in pool
- [x] Calculate: Amount of Token B that user owns in pool
- [x] Calculate: Share of Pool
- [x] "Create Pool" Button is visible

### 3. Create pool initiated
- [ ] Acknowledge transaction was initiated

### 4. Transaction Success Acknowledgement
- [ ] Acknowledge asynchronously successful transaction 

### 5. Transaction Failure Acknowledgement
- [ ] Acknowledge asynchronously failed transaction 

### 6. Backend Update
- [ ] Update liquidity pool values from backend recalculates frontend values

### 7 Insufficient Funds
- [x] Create pool button is not clickable
- [x] Display Insufficient funds message

### 8 No wallet connected
- [x] Can see Token Amounts they enter
- [x] Calculate Token A per B ratio
- [x] Calculate Token B per A ratio
- [x] Calculate Share of Pool
- [x] Create Pool button is not clickable

### 1. Appearance
- [x] Token A & Token B
- [x] Token A Amount & Token B Amount
- [x] Can see Token Amounts they enter
- [x] Calculate Token A per B ratio
- [x] Calculate Token B per A ratio
- [x] Calculate Share of Pool

### 2. Enter token values less than wallet values
- [x] Calculate Token A per B ratio
- [x] Calculate Token B per A ratio
- [x] Calculate Share of Pool
- [x] "Add liquidity" Button visible

### 3. Transaction Initiated Acknowledgement
- [ ] Acknowledge transaction initiated

### 4. Transaction Success Acknowledgement
- [ ] Acknowledge transaction succeeded

### 5. Transaction Error Acknowledgement
- [ ] Acknowledge transaction error/failed

### 6. Backend Update
- [ ] Share of pool updated by the backend (Code setup to do this reasonably easily but need info on exactly what API provides liquidity pool volume to calculate rate or stream interface etc.)

### 7 Insufficient Funds
- [x] Add Liquidity button is not clickable
- [x] Display Insufficient funds message

### 8 No wallet connected
- [x] Can see Token Amounts they enter
- [x] Calculate Token A per B ratio
- [x] Calculate Token B per A ratio
- [x] Calculate Share of Pool
- [x] Add Liquidity button is not clickable

### 1. Appearance
- [x] Token A & Token B
- [x] Token A Amount & Token B Amount
- [x] Calculate: % share of liquidity pool

### 2. Enter amounts less than / equal to prev amounts
- [ ] Amount
- [ ] Gas fees (How do we calculate these?)
- [ ] Remove button is visible

### 3. Transaction Initiated Acknowledgement
- [ ] Acknowledge transaction initiated

### 4. Transaction Success Acknowledgement
- [ ] Acknowledge asynchronously successful transaction 

### 5. Transaction Failure Acknowledgement
- [ ] Acknowledge asynchronously failed transaction 

### 6. Enter token values greater than amount added or greater that amount available in pool
- [ ] "Amount too high" message

### 7. Backend Update
- [ ] Share of pool updated by the backend

### 8. Wallet not connected 
- [x]  Remove LIquidity Button not available

### CODE REVIEW

File: ui/app/src/components/shared/CurrencyForm.vue [hunk 1]
```diff
@@ -0,0 +1,83 @@
+<template>
+   <div class="currency-form text--small">
+      <div class="input">
+        <SifInput :label="label" />
+      </div>
+      <div class="balance">
+        <span>Balance: 1200,7356</span>
```
Comment (59023g) [inline ui/app/src/components/shared/CurrencyForm.vue:7]: Is this supposed to be here?

File: ui/app/src/views/Ui.vue [hunk 2]
```diff
@@ -0,0 +1,239 @@
+<script lang="ts">
```
Comment (59023g) [inline ui/app/src/views/Ui.vue:1]: What's this file? cc @novelja

File: ui/core/scripts/getTopErcTokens.js [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+      },
+    ],
+    name: "Transfer",
+    type: "event",
+  },
+];
+
+const web3 = new Web3(
+  new Web3.providers.HttpProvider(
+    "https://mainnet.infura.io/v3/93cd052103fd44bd9cf855654e5804ac"
```
Comment (59023g) [inline ui/core/scripts/getTopErcTokens.js:230]: Why hardcode this?

File: ui/core/scripts/getTopErcTokens.js [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+      name,
+      symbol,
+      image,
+      contract_address,
+      asset_platform_id,
+      market_cap_rank,
+    })
+  );
+
+  fs.writeFileSync("data/erc20.json", JSON.stringify(erc20, null, 2));
```
Comment (59023g) [inline ui/core/scripts/getTopErcTokens.js:290]: Is this runtime?

File: ui/core/src/actions/ethWallet.ts [hunk 5]
```diff
@@ -8,6 +8,8 @@ export default ({
   api,
   store,
 }: ActionContext<"EthereumService", "wallet" | "asset">) => {
+  const etheriumState = api.EthereumService.getState();
```
Comment (59023g) [inline ui/core/src/actions/ethWallet.ts:11]: Typo

File: ui/core/src/entities/Pool.test.ts [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    name: "Rowan",
+    network: Network.SIFCHAIN,
+  });
+
+  // TODO: Confirm with Blockscience
+  // x = Sent Asset Amount, X = Sent Asset Pool Balance, Y = Received Asset Pool Balance
+
+  // Liquidity Fee = ( x^2 * Y ) / ( x + X )^2
+  // Trade Slip = x * (2*X + x) / (X * X)
+  // Swap Result = ( x * X * Y ) / ( x + X )^2
```
Comment (59023g) [inline ui/core/src/entities/Pool.test.ts:40]: Nice

File: ui/core/src/entities/formulae.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+export function calculateSwapResult(X: IFraction, x: IFraction, Y: IFraction) {
+  return x
+    .multiply(X)
+    .multiply(Y)
+    .divide(x.add(X).multiply(x.add(X)));
+}
+
+export function calculateExternalExternalSwapResult(
```
Comment (59023g) [inline ui/core/src/entities/formulae.ts:125]: Whoa

**Review by 59023g — DISMISSED** [review]
🚀

**Comment by 59023g** [comment]
Minor questions. Looks good. Ship it!

**Comment by ryardley** [comment]
@59023g I Got codecov happy perhaps you can review again?

**Comment by 59023g** [comment]
Looks good. I created Asana tasks for the {Add Liquidity, Remove Liquidity, Create Pool, Swap} Notifications. And other tasks without check marks

**Review by 59023g — APPROVED** [review]

**Pull request merged**